src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -1795,25 +1802,25 @@ private Symbol RemapSymbolIfNecessary(Symbol symbol)
         /// </summary>
         internal abstract Symbol RemapSymbolIfNecessaryCore(Symbol symbol);
 
-        private static ImmutableArray<ISymbol> FilterNotReferencable(ImmutableArray<ISymbol> sealedResults)
+        private static ImmutableArray<ISymbol> FilterNotReferenceable(SegmentedList<ISymbol> sealedResults)
         {
-            ArrayBuilder<ISymbol> builder = null;
-            int pos = 0;
-            foreach (var result in sealedResults)
+            var cannotBeReferencedByNameCount = sealedResults.Count(static s => !s.CanBeReferencedByName);
+
+            if (cannotBeReferencedByNameCount == 0)
             {
-                if (result.CanBeReferencedByName)
-                {
-                    builder?.Add(result);
-                }
-                else if (builder == null)
+                return sealedResults.AsImmutable();
+            }
+
+            var builder = ImmutableArray.CreateBuilder<ISymbol>(sealedResults.Count - cannotBeReferencedByNameCount);
```

Comment:
Why wouldn't we want to use `ArrayBuilder` here? Either the size is compatible with the pool, and we'll avoid a new allocation, or we will get a new builder that will then get discarded?

Reply:
There are three possibilities the way I see it:

1) ArrayBuilder finds an array that is too small and internally resizes it's builder to the new capacity. This allocates on retrieval but not on ToImmutableAndFree.
2) ArrayBuilder finds an array exactly the right size. This doesn't allocate upon retrieval or on ToImmutableAndFree.
3) ArrayBuilder finds an array that exceeds the requested size. This doesn't allocate upon retrieval but does on ToImmutableAndFree.

The only case where we allocate less is when the retrieved array is exactly the right size, but then we lost that entry from the pool. And all three cases require the ArrayBuilder to search it's pool for an available item.

I can switch to that if you prefer, but I don't believe it provides any performance benefit while needing to do the (slight) overhead of searching the pool for an available item.

Reply:
Hmm. I think I misunderstood how ArrayBuilder's GetInstance works then. Let's leave a comment as to the reason that we aren't using it here.